日報の言及機能を作る

gaki273 opened the pull request.

itagakisan committed:
`Mention`モデル追加とマイグレーション実行 Model追加

[目的]
- 日報の言及機能を実装する

[変更内容]
- `Mention`モデルを生成
- `mentions`テーブルを作成（`mentioning_id`、`mentioned_id`）
- ユニーク複合インデックスを追加
- スキーマを更新

itagakisan committed:
`Mention`モデルのアソシエーション追加 機能追加

[目的]
- 日報間のメンション関係を定義する

[変更内容]
- `Mention`モデルに`belongs_to`アソシエーションを追加
- `Report`モデルに`has_many`アソシエーションを追加
- `active_mentions`と`passive_mentions`でメンション関係を定義

itagakisan committed:
メンションの表示を行うパーシャルを追加 View追加

[目的]
- 日報ページにメンション情報を表示する

[変更内容]
- 日報の詳細ページにメンションパーシャルを追加
- `shared/_mentions.html.erb`パーシャルを追加

itagakisan committed:
`Report`モデルにメンション関連アソシエーション追加 機能追加

[目的]
- 日報間のメンション関係に直接アクセスできるようにする

[変更内容]
- `mentioning_reports`アソシエーションを追加
- `mentioned_reports`アソシエーションを追加

itagakisan committed:
言及（メンション）自動検出する`update_mentions`メソッドを追加  機能追加

[目的]
- 日報内容のリンクからメンションを自動作成する

[変更内容]
- `update_mentions`メソッドを実装
- 正規表現で日報URLを抽出して`Mention`レコードを作成
- `after_save`コールバックで`update_mentions`を実行

itagakisan committed:
メンション表示の実装 機能追加

[目的]
- その日報を言及している他の日報を画面に表示する

[変更内容]
- 条件分岐を`Mention.order(:id).includes(:user).presence`から`@report.mentioned_reports&.any?`に変更
- `mentioned_reports.each`でループ処理を実装
- `report.user.name_or_email`でユーザー名を表示
- `link_to report.title, report_path(report.id)`でタイトルリンクを表示

itagakisan committed:
メンション機能の日本語対応追加

[目的]
- メンション表示機能で使用するテキストを日本語化

[変更内容]
- `ja.yml`に`views.shared.mentions.no_mentions`を追加
- `translation_ja.yml`に`activerecord.models.mention`を追加

yoshitsugu commented on app/models/report.rb:
```diff
@@ -4,14 +4,30 @@ class Report < ApplicationRecord
     created_at.to_date
   end
+
+  def update_mentions
```
内部的に `create!` を使っているので、 Railsの命名規則的にはここも `update_mentions!` としたほうがよさそうです。

gaki273 replied:
かしこまりました！
修正いたします。

yoshitsugu commented on app/models/report.rb:
```diff
@@ -4,14 +4,30 @@ class Report < ApplicationRecord
   validates :title, presence: true
   validates :content, presence: true
 
+  after_save :update_mentions
```
**今回はこれでOKです。**
after_saveなどのコールバックと呼ばれるものはコードを読む側からすると処理が追いづらくなるため、素直にcontrollerなどから呼びだす場合と比べてデバッグのときなどに難易度があがります。
また、複数のコールバックを定義してしまうと、定義順が実行順となりますので、あるコールバックで変わった内容を想定しながら次のコールバックを書かないといけない、などの認知負荷があがる原因にもなります。そのため、コールバックで書かないとコードのメンテナンスコストがかなり上がってしまいそうな場合など明確なデメリットがない限りは素直にcontrollerなどから呼び出す形式のほうがわかりやすいコードになることが多いです。

同様の話としてcontrollerのbefore_actionの話があります。
参考: https://speakerdeck.com/shinkufencer/how-to-using-before-action-with-happy-in-rails
こちらも参考までに。

yoshitsugu replied:
ちなみに、学習のための質問ですが、

> 日報の保存には成功したが、言及データの保存には失敗している、という状況が発生しないようにする

という注意があります。これはコールバックによって達成されますか？達成される場合、どういう仕組みでそうなるのかわかりますか？

gaki273 replied:
コメントいただきありがとうございます。コールバックを多用することは注意が必要なのですね。以後、使用に気を付けて、明確なデメリットがない限りは使用を控えるようにしてまいります。

トランザクションに関してですが、今回のコールバックでも達成されていると考えます。
> 日報の保存には成功したが、言及データの保存には失敗している、という状況

というのは
`after_save`コールバック内で`create!`メソッドで例外が発生したときだと思います。
`create!`メソッドはバリデーションに失敗した場合、例外を発生させます。`create!`メソッドで例外が発生した場合、トランザクションはロールバックされ、データベースへの変更は取り消されます。`after_save`はトランザクション内で実行されるため、コールバック内でエラーが発生した場合、トランザクション全体がロールバックされるため、今回の実装では上記の注意事項を考慮できていると考えております。
しかし、先のコメントでもご指摘があったように命名規則の観点で、`update_mentions!`と修正したほうが良いと考えます。また、コードの可読性を考慮したり、今後の実装を考えると、ご指摘のとおり素直にcontrollerなどから呼び出す形式のほうが良いと考えます。

yoshitsugu replied:
> after_saveはトランザクション内で実行される

特にここがポイントですね。OKです :+1:

yoshitsugu commented on db/migrate/20250801023447_create_mentions.rb:
```diff
@@ -0,0 +1,11 @@
+class CreateMentions < ActiveRecord::Migration[7.0]
+  def change
+    create_table :mentions do |t|
+      t.integer :mentioning_id
+      t.integer :mentioned_id
```
`integer` ではなく専用の `references` を使うとよいです。一般的にsqlite以外のDBではidはbigint型になるので、 `integer` だと足りないです。
https://railsguides.jp/active_record_migrations.html#%E9%96%A2%E9%80%A3%E4%BB%98%E3%81%91%E3%82%92%E4%BD%9C%E6%88%90%E3%81%99%E3%82%8B
外部キーなども検討してください。
また、必須のカラムは `null: false` になるように注意しましょう。

gaki273 replied:
DBによってidの方が異なるのですね。
かしこまりました。
上記のコメントを考慮して修正いたします！

yoshitsugu reviewed (commented):
コメントしました！

itagakisan committed:
`update_mentions`メソッドを`update_mentions!`メソッドに変更

itagakisan committed:
データベーススキーマを更新 修正

[目的]
- データベースの整合性とパフォーマンスを向上させる

[変更内容]
- mentionsテーブルにnull: false制約を追加
- mentionsテーブルに外部キー制約を追加（reportsテーブル参照）
- mentionsテーブルに個別インデックスを追加
- Active Storageの型をintegerからbigintに変更

yoshitsugu commented on app/models/report.rb:
```diff
@@ -4,14 +4,30 @@ class Report < ApplicationRecord
+
+  def update_mentions!
+    active_mentions.destroy_all
+    report_ids = content.scan(%r{http://127\.0\.0\.1:3000/reports/(\d+)}).flatten.map(&:to_i).uniq
```
前回気づいていなくても申し訳ないですが、

>  検知すべきドメイン（host）は http://localhost:3000/ だけでOKです

という注意がプラクティスにあり、 `127.0.0.1` でも間違ってないのですが、要件上は `localhost` を検知することが求められています。要件に沿うように修正おねがいします。

gaki273 replied:
かしこまりました！
私も気づいていていませんでした💦
修正いたします！

yoshitsugu reviewed (commented):
一点細かいですがコメントしました！

itagakisan committed:
メンション抽出のURLパターンをlocalhostに変更 修正

[目的]
- 必要要件に合わせる

[変更内容]
- 正規表現パターンを127.0.0.1からlocalhostに変更

yoshitsugu reviewed (approved):
OKです！
